### ISSUE TITLE
Can @nnx.custom_vjp bwd function update its arguments?

### ISSUE THREAD

**Issue by hrbigelow** [google/flax#4344]
Hi @cgarciae,

Is it possible for the @nnx.custom_vjp backward function to update its arguments?

I noticed your comment [here](https://github.com/google/flax/blob/3e17c0663a16985359527319be002e08cb0a06cd/flax/nnx/transforms/autodiff.py#L379-L381) which suggests it is not supposed to:

```python
# 4. BwdFn: wraps the user's bwd function, it converts its input pytrees to graph nodes
#    and output graph nodes to pytrees. It doesn't need to be aware of the outer context
#    since it will never update the outer references as it runs during the backward pass.
```

But, that would be ashame because it would be a powerful feature to be able to perform parameter updates during the backward pass.

At the moment I am trying to implement a so-called "immediately updated gradient" which is an unconventional training algorithm but potentially faster converging.  Instead of computing the gradient of the entire model parameters in one backward pass, it computes each layer's parameter gradients, then applies them in-place, then computes model input gradients based on updated parameters.  The math is described in this [jax discussion](https://github.com/jax-ml/jax/discussions/24265).

I tried the following but the update `opt.update(opt_g['model'])` does not appear to be executing (or perhaps it is but somehow being performed on a copy which is then thrown away?)

Any thoughts or workarounds would be much appreciated!

The full experiment is [here](https://github.com/hrbigelow/transformer-aiayn/blob/flax-port/tests/mnist_nnx_custom_vjp.py)

```python
model = nnx.Module(...)
tx = optax.adamw(0.005)
opt = nnx.Optimizer(model, tx)

def make_custom_vjp(opt):
  def fn(opt, x):
    return opt.model(x)

  @nnx.custom_vjp
  def primal(opt, x):
    return fn(opt, x)

  def fn_fwd(opt, x):
    return fn(opt, x), (opt, x)

  def fn_bwd(res, g):
    ins_g, out_g = g
    opt, x = res
    opt_g = nnx_vjp(lambda opt_arg: opt_arg.model(x), opt, out_g)
    # nnx.display(opt_g)
    # param_g = jax.lax.pmean(param_g, 'dev') # ? each gradient on a different device
    opt.update(opt_g['model']) 
    data_g = nnx_vjp(lambda x_arg: fn(opt, x_arg), x, out_g)
    # nnx.display(data_g)
    # opt_pars = nnx.state(opt)
    return (opt_g, data_g) # TODO: param_g unnecessary

  primal.defvjp(fn_fwd, fn_bwd)
  return primal
```

**Comment by cgarciae** [comment]
@hrbigelow sadly this is currently more of a limitation of JAX: `bwd` can only specify tangents but cannot create/update state. To get around this @mattjj has been working on a [Mutable Array](https://github.com/jax-ml/jax/blob/main/tests/mutable_array_test.py) reference that can be updated during the backward pass.

Ideally you could just store a Mutable Array inside Variable and just perform the updates. However there might be some issues as you cannot return Mutable Arrays from jax transforms, so we might have to find a way to add support for them in NNX.

**Comment by hrbigelow** [comment]
@cgarciae thanks" you, that makes sense.  Yeah it would be a very valuable addition to flax / jax to be able to do this.  I'll close this for now but keep an eye on Matt's mutablearray effort.
